pkg/pubsub/pubsub_init.go

```diff
@@ -0,0 +1,95 @@
+package pubsub
+
+import (
+	"context"
+	"fmt"
+
+	psClient "github.com/open-policy-agent/gatekeeper/pkg/pubsub/client"
+	"github.com/open-policy-agent/gatekeeper/pkg/util"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubeClient "sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+var log = logf.Log.WithName("pubsub")
+
+var _ manager.Runnable = &runner{}
+
+type runner struct {
+	mgr manager.Manager
+}
+
+func AddToManager(m manager.Manager) error {
+	mr := new(m)
+	return m.Add(mr)
+}
+
+func new(mgr manager.Manager) *runner {
+	mr := &runner{
+		mgr: mgr,
+	}
+	return mr
+}
+
+var initiatedTools []psClient.PubSub
```

Comment:
This should be a map if we are going to route publishes to particular clients.

Reply:
Maybe this might be more suited for a follow-up PR? cause IMO routing publishes to particular clients requires a control mechanism as well that allows users to select routing preference and then logic to actually make code abide by the routing mechanism. 

Wondering if we want to think on the same plane in this change or next?

Reply:
IIRC we talked about this when we met.

Routing to a specific endpoint would be done by a `(connection, topic)` tuple (naming bikeshedding notwithstanding)

for example: `Publish("infra-SQS", "alerts", "bad thing is happening right now")`

If you recall, we looked at how Dapr only had a single client, and created the concept of a "component":

https://docs.dapr.io/developing-applications/building-blocks/pubsub/howto-publish-subscribe/#set-up-the-pubsub-component

Each new Dapr "connection" is really just calling Dapr for a specific component (which we can fake by creating a dummy "connection" object that hardcodes the component name so we can treat multiple Dapr registrations the same as multiple RabbitMQ registrations).

IMO figuring this stuff out is in scope because the point of designing an interface is to make sure you have something that meets forseeable requirements.

Doing this work is what will let users use Dapr or RabbitMQ or whatever works with their infrastructure. 

Designing the interface to meet forseeable requirements is what allows us to minimize the risk re-writing code (and, worse, user-facing config), once we hit the next stage of implementation.

Assuming this function signature: `Publish("infra-SQS", "alerts", "bad thing is happening right now")`, the keys to the map would be the first argument to `Publish()`. The routing code would be (adding context b/c we probably should and locking b/c I assume this will need to be thread-safe):

```golang
def (p *PubSubRouter) Publish(ctx context.Context, connection, topic string, msg interface{}) err {
   p.mux.RLock()
   defer p.mux.RUnlock()
   c, ok := p.connections[connection]
   if !ok {
      return fmt.Errorf("No connection named %q", connection)
   }
   return c.Publish(ctx, topic, msg)
}
```

Which I'm guessing would not look very different from what you would have done had this been a list.

Do you see how getting the details right for future uses lessens risk of extra future work and/or need to break backwards compatibility without much higher coding costs (if any)?

Reply:
@maxsmythe This solves the problem of routing, as the user can pass the connection (name of the connection) in the `Publish` to determine where to publish the message. 

Let's say we right now make call to `Publish` from audit manager to publish audit violations using `Publish(ctx context.Context, connection, topic string, msg interface{})`
How do we take user input on which connection name to use [here](https://github.com/open-policy-agent/gatekeeper/pull/2538/files#diff-2a233457fc5c89cf0dd91c044d7582aa1cc0c835feabf4638882ac54ba3ca63aR762)?

Reply:
The audit case is trivial... right now we are saying: "let's let users configure audit to push rejections to one place", so whatever knobs we're giving the user (flags, config file, CRD, ConfigMap, etc.) should require they specify both the connection and the topic.

Reply:
I am still working on an efficient way to allow users to input this information and make logical decisions in the code.

Reply:
It may be sufficient to just define how the data is passed in a way that makes it not-the-interface's-problem.

For example, what if we had two methods (I'm using `Connection` to refer to the interface for "an object that knows how to send messages to a given pubsub queue", I think you call this `PubSub`... connection seems easier to talk about since "PubSub" is an overloaded term):

```golang
// note I have not added error handling or context management

AddConnection(connection Connection)

RemoveConnection(connection Connection)

// For use cases where a controller is handling configuration

// Reconcile watches some config CRD, ConfigMaps, Secrets, whatever
def (r *reconciler) Reconcile(...) {
   config := client.Get(something) // load in the appropriate K8s object
   // because users can write whatever reconciler the actual method/signature required to
   // create a connection is irrelevant to how hard it is to fork (see mutator controllers for how
   // this can be made generic

   // an added bonus is a controller can retrieve a Secret separately and inject it into the config
   connection := newConnectionForConfig(config)
   r.pubSub.AddConnection(config)

   // RemoveConnection() handles when the config resource is deleted
}

// Users who want to support additional kinds could add a new controller
// and/or we could also try to design a "generic" controller that can handle
// the most common use cases

// simplified example of loading config from local files or flags at startup time (no reconciliation)


// flags
var (
  // flags doesn't work so well for the multiple connection use case, which
  // isn't really an interest for the audit MVP I don't think
  connectionCert = flags.String()
  connectionHost = flags.String()
  connectionTopic = flags.String()
)

def main() {
   // note that this doesn't play as cleanly with multiple possible connections, as
   // the constructor shape is coupled to the flags, meaning forks may be more heavy weight
   connection := newConnection(connectionCert, connectionHost, connectionTopic)
   pubSub.AddConnection(connection)
}

// using a file
def main() {
   // readFile is a function that knows how to read a generic config wrapper, example yaml:
   // driver: "rabbitMQ" # required, a string that identifies the driver used to load the config
   // config: <random yaml> # the contents of this field are read by a driver-specific parser
  //                                             # probably has the same shape as an object watched by the controller
   config := readFile(configFile)

   
   driver := config.Driver()

   // dynamically load the parser code for the appropriate config, run it
   // and use that to create the driver. There are a couple of ways to divide
   // this functionality, this is just an example
   driverConfig := pubsubRegistry.ConfigParser(driver).Parse(config)
   connection := pubsubRegistry.Connection(driver, driverConfig)
   pubSub.AddConnection(connection)

}
```

From the above, I think a key technique for minimizing forking is to be able to parse configs dynamically.